admin-dev/themes/new-theme/js/components/grid/extension/async-submit-bulk-action-extension.js

```diff
@@ -0,0 +1,162 @@
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+const $ = window.$;
+
+export default class AsyncSubmitBulkActionExtension {
+  constructor() {
+    return {
+      extend: grid => this.extend(grid),
+    };
+  }
+
+  /**
+   * Extend grid with bulk action submitting
+   *
+   * @param {Grid} grid
+   */
+  extend(grid) {
+    grid.getContainer().on('click', '.js-bulk-action-submit-async-btn', (event) => {
+      this.submit(event, grid);
+    });
+  }
+
+  /**
+   * Collects required data and submits the form.
+   * @param event
+   * @param grid
+   */
+  submit(event, grid) {
+    const $submitBtn = $(event.currentTarget);
+    const confirmMessage = $submitBtn.data('confirm-message');
+
+    if (typeof confirmMessage !== 'undefined' && confirmMessage.length > 0 && !confirm(confirmMessage)) {
+      return;
+    }
+
+    const $form = $(`#${grid.getId()}_filter_form`);
+    const $checkedInputs = this.getCheckedInputs($form);
+    const ids = this.getCheckedIds($checkedInputs);
+
+    const inputName = this.getInputName($checkedInputs);
+    const chunkSize = parseInt($submitBtn.data('step'), 10);
+    const chunkedIds = this.chunkArray(ids, chunkSize);
+    const submitMethod = $submitBtn.data('form-method');
+    const formUrl = $submitBtn.data('form-url');
+
+    if (!['POST', 'GET', 'DELETE', 'PUT'].includes(submitMethod)) {
+      return;
+    }
+
+    /**
+     * Async submit action which repeats itself in recursive order.
+     * @param items
+     * @param successCallback
+     * @param errorCallback
+     */
+    const asyncSubmit = (items, successCallback, errorCallback) => {
```

Comment:
you could define this as a class function instead of a local const

Reply:
if I do that there will be one disadvantage:

I will need to pass two more variables `submitMethod` and `formUrl` via method params or I will need to define them as private properties in the class to be able to use them. None of the options looks clean for me.

Im taking here advantage of having anonymous function here. Without using it I would need to think of how to pass these two extra variables which dont participate in recursive iterations

Reply:
I would be fine with using private fields, but I get your point of view And it's valid so nevermind what I said :)